File: go/search/binary_search.go
package main

//二分查找

//迭代实现
func binarySearch1(data []int, target int) int {
	low, high := 0, len(data)-1
	for low <= high {
		mid := low + (high-low)/2
		if target > data[mid] {
			low = mid + 1
		} else if target < data[mid] {
			high = mid - 1
		} else {
			return mid
		}
	}
	return -1
}

//递归实现
func binarySearch2(data []int, target int) int {
	return helper(data, target, 0, len(data)-1)
}

func helper(data []int, target int, low int, high int) int {
	if low <= high {
		mid := low + (high-low)/2
		if data[mid] > target {
			return helper(data, target, low, mid-1)
		} else if data[mid] < target {
			return helper(data, target, mid+1, high)
		} else {
			return mid
		}
	}
	return -1
}

File: cpp/search/binary_search.cpp
#include "iostream"
#include "vector"

using namespace std;

class BinarySearch {
public:
    // 迭代
    int search(vector<int>& nums, int target) {
        int low = 0;
        int high = nums.size() - 1;
        while(low <= high) {
            int mid = low + (high - low) / 2;
            if(nums[mid] > target) {
                high = mid - 1;
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;
    }

    // 递归
    int search_recursive(vector<int>& nums, int target) {
        return helper(nums, target, 0, nums.size() - 1);
    }

    int helper(vector<int>& nums, int target, int low, int high) {
        if (low > high) {
            return -1;
        }

        int mid = low + (high - low) / 2;
        if(nums[mid] > target) {
            return helper(nums, target, low, mid - 1);
        }
        if(nums[mid] < target) {
            return helper(nums, target, mid + 1, high);
        }
        return mid;
    }
};
